### DISCUSSION TITLE
njsPC not controlling chlorinator?

### DISCUSSION THREAD

**Discussion by TopJim** [tagyoureit/nodejs-poolController#896] [RS485] [answered]
njsPC does not appear to be controlling my IC20. DashPanel reports "Chlorinator Comms Error" but in "Comms" I am receiving data and I am able to change IC20 output and see salt ppm. Replay attached.

Background: I recently set up a chemical controller on Raspi 4b with njsPC/REM and Intellicenter i10d. My pool is closed for the winter; so right now it's only controlling my spa (although I do have all EZOs for pool set up in REM).
- Spa IC20 is on Aux USB1 RS-485, Intellicenter Aux3 relay. I have spa and IC20 on 24 hr schedules. IC20 set to 1%.
- intellicenter is on USB0 RS-485 (and when I get the pool running, its IC40 will also be on this bus).
- Other equipment: 2 ea EZO pH, ORP, RTD, PRS on Whitebox Labs T3 (1 HAT for pool, 1 for spa). Sequent 4rel4ind board  controlling 2 ea Stenner E10FXA 24VDC.
[replay.zip](https://github.com/tagyoureit/nodejs-poolController/files/14212459/replay.zip)

**Comment by rstrouse** [comment]
Is the spa the second body or the first.   I think it is checking the chlorinator for the first body incorrectly.   If it is then I will make sure it looks at the status for the correct body.

**Comment by TopJim** [comment]
Spa is set as 2nd body. Pool is off for the winter, so its pump and IC40 not powered. Davey JonesOn Feb 8, 2024, at 22:58, rstrouse ***@***.***> wrote:﻿
Is the spa the second body or the first.   I think it is checking the chlorinator for the first body incorrectly.   If it is then I will make sure it looks at the status for the correct body.

**Comment by rstrouse** [comment]
Yep it is not taking the second body into consideration when checking to see if the chlorinator is alive.  As a result it displays a warning that the chlorinator is not communicating.  I'll take a look and change this so it looks at only chlorinators on the same body as REM.

**Comment by TopJim** [comment]
Thanks for checking!

**Comment by TopJim** [comment]
I don't know if this is part of the issue or totally separate but I'm also getting similar behavior for the spa stenner pump relay. I am able to turn on the stenner pump using the "dose" button or the calibration function, but an ORP setpoint will not turn on the pump relay. Example: ORP is reading 750, my setpoint is 800--nothing happens.

**Reply by rstrouse** [reply]
So you have a chlorinator and a chlorine tank on the same body?

**Reply by TopJim** [reply]
I'm sorry, I made an error. I only have a chlorinator. I didn't mean to say ORP, but rather pH. So to restate:
I am able to turn on the stenner pump using the "dose" button or the calibration function, but an pH setpoint will not turn on the pump relay. Example: pH is 7.9, my setpoint is 7.5--nothing happens.

**Reply by rstrouse** [reply]
Do you have it set to dose by manual?  If so you need to set up a dose by volume.

**Reply by TopJim** [reply]
I have it set by volume.
I just tried something different and now the Spa auto-dosing works: I deleted the Pool IC40 and the Pool RemChem. So maybe there's some interaction like with the chlorinators on 2 bodies (chlorinator still has "comms error").
Now I go back and try to add the pool IC40 and get an error:
ApiError: You may not install a chlorinator on an intellicenter system that is assigned to the Primary Port that is under Nixe control
    at NixieChlorinator.setChlorinatorAsync (/home/pi/nodejs-poolController/controller/nixie/chemistry/Chlorinator.ts:130:104)
    at NixieChlorinatorCollection.setChlorinatorAsync (/home/pi/nodejs-poolController/controller/nixie/chemistry/Chlorinator.ts:44:25)
    at IntelliCenterChlorinatorCommands.setChlorAsync (/home/pi/nodejs-poolController/controller/boards/SystemBoard.ts:3364:40)
    at IntelliCenterChlorinatorCommands.setChlorAsync (/home/pi/nodejs-poolController/controller/boards/IntelliCenterBoard.ts:2376:95)
    at /home/pi/nodejs-poolController/web/services/config/Config.ts:668:57
    at Layer.handle [as handle_request] (/home/pi/nodejs-poolController/node_modules/express/lib/router/layer.js:95:5)
    at next (/home/pi/nodejs-poolController/node_modules/express/lib/router/route.js:144:13)
    at Route.dispatch (/home/pi/nodejs-poolController/node_modules/express/lib/router/route.js:114:3)
    at Layer.handle [as handle_request] (/home/pi/nodejs-poolController/node_modules/express/lib/router/layer.js:95:5)
    at /home/pi/nodejs-poolController/node_modules/express/lib/router/index.js:284:15
    at Function.process_params (/home/pi/nodejs-poolController/node_modules/express/lib/router/index.js:346:12)
    at next (/home/pi/nodejs-poolController/node_modules/express/lib/router/index.js:280:10)
    at /home/pi/nodejs-poolController/web/Server.ts:749:25
    at Layer.handle [as handle_request] (/home/pi/nodejs-poolController/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/home/pi/nodejs-poolController/node_modules/express/lib/router/index.js:328:13)
    at /home/pi/nodejs-poolController/node_modules/express/lib/router/index.js:286:9

**Reply by rstrouse** [reply]
That is correct.  The IntelliCenter has to be in control of the IC40 on the primary port.   Do not select Nixie to be in control.

**Reply by TopJim** [reply]
Then I deleted both pool IC40 and spa IC20 from DashPanel Chemistry tab and re-added them in order. And then re-configured the RemChem for Spa in DashPanel The auto-dose is now working. Don't know what the issue was at first. Still have the chlorinator comms error, though.

**Reply by rstrouse** [reply]
Pull njsPC it was in fact looking at the wrong chlorinator for dosing.

**Comment by TopJim** [comment]
I gotta go somewhere so will pull in about 2 hrs. I need to ask a moronic question tho: how do I “pull”?Davey JonesOn Feb 10, 2024, at 16:06, rstrouse ***@***.***> wrote:﻿
Pull njsPC it was in fact looking at the wrong chlorinator for dosing.

**Reply by rstrouse** [reply]
On the pi for the pool controller

`cd ~/nodejs-poolController`
`git pull`
`npm i`

**Reply by TopJim** [reply]
Do I need to stop njsPC  or should I stop all 3 (njsPC, dashpanel, REM) before pulling?

**Reply by rstrouse** [reply]
If you are using pm2 it will automatically stop then reload when you are done.

**Reply by TopJim** [reply]
I did a pull (thx for the help on that), but it's not working now. On DashPanel I'm getting an error:
Error: connect ECONNREFUSED 10.0.0.26:4200
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1555:16)
I powered everything down and restarted, but error persists. In "settings" it shows dashPanel 8.0.0
git branch master
git commit e3ff402

**Comment by TopJim** [comment]
Also, I search for server and it says:
"No running nodesjs-PoolController servers could be found. Enable SSDP on the pool controller application configuration"

**Reply by TopJim** [reply]
maybe I shouldn't have done the npm audit fix? Could that have screwed it up?

**Comment by rstrouse** [comment]
Never ever do that.   I hate that message.  Do a pull again and if you get an error run git reset —hard and pull again.

**Comment by TopJim** [comment]
OK I restored and pulled again. Git commit for njsPC 8.0.1 is c3f9a06
It still says "chlorinator comms error"
It says "Chlorinating .01458 -- 15 min" but it shouldn't be because my ORP is 798 and my setpoint is 750. So it appears it's not being controlled because output is still showing 1%.

I have a chemical controller set up for the pool (but its IC40 is off because the pump CB is off; Intellicenter think the pump is on). 

Don't know if it means anything or not, but on DashPanel "REMChemPool" is showing pH and ORP nearly equal to "REMChemSpa" even though REM is reporting Pool EZOs as 678 mV & 13.9 ph (because I don't have sensors hooked up).

I checked the feeds (that I am able to configure). Spa is controller ID 51, Pool is 50. Replay attached.
[replay.zip](https://github.com/tagyoureit/nodejs-poolController/files/14231126/replay.zip)

**Reply by TopJim** [reply]
Feeds are for Spa:
Ezo/connection/val/to
RTD/internal/tempC/ph_spa
PRS/none
ORP/david/ORP/chemcontroller 51
pH/david/ph/chemcontroller 51

For pool:
RTD/internal/tempC/ph_pool
PRS/none
ORP/david/orp/chemcontroller 50
pH/david/ph/chemcontroller 50

Is that all what it should be? Is the controller ID a arbitrary number? Otherwise, IDK why it’s 50 and 51.

**Reply by rstrouse** [reply]
It is 50+ because IntelliChem has to be 1.  Look in the poolConfig.json but that seems right.

**Reply by rstrouse** [reply]
I’ll have a look at your replay tomorrow.

**Comment by TopJim** [comment]
For the time being, I switched the Spa IC20 to the intellicenter RS-485. I figured since I use the pool 7 months out of the year, would make better sense to have the Spa IC20 on the main data bus. All appears well for now. We'll see what happens when I get the pool going in Apr/Mar timeframe.
One question: I have the Spa IC20 set to 1% with dynamic dosing on, so REMChem cycles it from 0% to 1%. Should I set a higher percentage setting?

**Reply by tagyoureit** [reply]
The IC20 (and all models) only ever are on or off.  They don't actually run at a percentage.  1% is the amount of time that the salt cell will be on.  AKA it will be on full (100%) for 1% of the day.  If your chlorine levels are where you want them, that's the best indicator.  You can also use your ORP feed as a target and the chlorinator timing will adjust to best meet that target.

**Reply by TopJim** [reply]
Cool, so just that I understand fully (and correct me if I'm wrong): when dynamic dosing is enabled the % setting has no meaning.
Thanks again to you and to rstrouse for helping me and authoring the software. It's pretty sweet.

Next project is to be able to bring up DashPanel outside my LAN, such as when I'm traveling away from home.
What's the best way to do this? I'm not that knowledgeable when it comes to networking, so the only thing that comes to mind is getting a static IP from my ISP, but since that costs a monthly fee, I'd like to do something else, if possible.

**Reply by TopJim** [reply]
Nevermind, I discovered ZeroTier. Works great.

**Comment by TopJim** [comment]
I now have my pool and spa running. I installed the chlorinators in DashPanel like this:
Pool --> i10d --> primary USB port 0 (RemChem refers to controller ID 50).
Spa --> external --> Aux USB port 1 (RemChem refers to Controller ID 51)

Poolconfig.json shows pool IC40 on ID 1 and Spa IC20 on id 50. DashPanel reports "Controller Body Mismatch" under "Pool RemChem". I tried changing RemChem controller ID's to match 1 and 50, but still get "Controller Body Mismatch" and in DashPanel both Pool & Spa are reporting values from Spa ORP sensor. So I switched it back to 50 and 51 in RemChem.

Some other symptoms: 
1. I am able to manually control Spa IC20 from DashPanel.
2. I am seemingly able to control Pool IC40 from DashPanel, but when I look at the IC40 panel it seems that it is being interrupted. The indicators lights do not show steady but briefly flash at 5 sec or so intervals.
3.  Somewhat similar behavior when I choose dynamic for pool: it sort of flashes 100% at odd intervals. It's almost like njsPC and Intellicenter are fighting for control of the Pool IC40. I imagine: NjsPC will tell it to go to 100% and then Intellicenter will tell it to go back to 0 (or whatever intellicenter is set to) and repeats.
4. When I choose dynamic for spa, it doesn't control the spa IC20, but it controls the Pool IC40 with the same behavior as the above.
5. I tried installing the Pool IC40 as "external", but when I then try to install the Spa IC20 on "external" get an error message "You may not install a chlorinator on an intellicenter system that is assigned to the Primary Port that is under Nixe control" and then it deletes the pool IC40. Sometimes I don't get an error message, but it deletes the Pool IC40 whenever I try this.

Is there a bug which is coding the controller ID's in the chlorinator section of poolConfig.json incorrectly?
Or am I doing something wrong?
[poolConfig.json](https://github.com/tagyoureit/nodejs-poolController/files/14841754/poolConfig.json)

**Reply by tagyoureit** [reply]
njsPC and Intellicenter can't both be controlling the chlorinator.  REMChem on the pool body can be used to control your stenner but not the chlorinator (at least not when it is connected to Intellicenter).

**Reply by TopJim** [reply]
Is the solution to get another RS-485 USB dongle and have intellicenter on USB 0, Pool IC40 on USB 1, and spa IC20 on USB 2? Is this possible?

**Reply by tagyoureit** [reply]
Yes, that will work.  The chlorinator and REM Chem both get associated with a body which links the two.  And the chlorinator will talk on whichever port it is assigned to.

**Reply by TopJim** [reply]
Thx. I’ll check it out and see what happens.

**Reply by TopJim** [reply]
I connected:
USB 0: i10d intellicenter
USB 1: IC20 (spa)
USB 2: IC40 (pool)
I can't seem to make it work. Very briefly, on one of my iterations, I was able to manually and dynamically control the IC40, and manually control the IC20--but not dynamically control the IC20.
Can you give me some direction on how to set up the intellichlors so that RemChem dynamically controls them both?
To be specific: when I set up the first intellichlor, let's say the IC40, I have an option to select i10d, nixie, or external. And then the 2nd, IC20, I have a nixie and external option. What's the proper thing to select and the sequence?

**Reply by tagyoureit** [reply]
The IC20 should be set up as a Nixie chlorinator and set to the spa body and USB1.  
The IC40 should be set up as a Nixie chlorinator and set to the pool body and USB2.
This will allow Nixie to independently control both chlorinators.

In each of your 2 REM Chem controllers, you'll select the body (pool or spa) and enable ORP and select the `Use Chlorinator` option and dose by `Dynamic base on ORP Setpoint`.
![image](https://github.com/tagyoureit/nodejs-poolController/assets/1646840/9dea835f-1531-4204-9560-2ff95e2f0966)

**Reply by TopJim** [reply]
Thx. I’ll check it and report back. Might be a day or two. Got a bit of good weather at the moment and taking the opportunity to finish up my re-piping project. Gotta paint my pipes next and finish re-enabling a waterfall. Also I think one of my ORP probes must be bad. It’s been calibrated a couple times, but even with a fc of 8 it indicates at the moment 129. When chlorinating it goes down. Obviously it’s wrong bc the other body with pretty much the same chemistry is reporting 700.

**Reply by tagyoureit** [reply]
Try to switch the probes/EZO circuits and see if you get the same results.  Helps to eliminate any wiring or other oddities.  But the probes do go bad.  If you need to buy a new one, lots of ppl have had really good luck with the industrial probes.  They cost a bit more, but seems to last a lot longer.

**Reply by TopJim** [reply]
I set it up like you said and the pool IC40 is working, but not the spa IC20.
Symptoms:
- SpaREMChem says it's chlorinating, but Spa IC20 says output 0%, and the IC20 physical panel shows 0%.
- When I turn off the pool IC40 via relay, SpaRemChem reports a "chlorinator comms error" for the spa IC20. When I turn the pool IC40 back on, the "chlorinator comms error" on the spa goes away. However, on my connections tab, the spa IC20 reports 0 failures.

Replay attached.
[replay (1).zip](https://github.com/tagyoureit/nodejs-poolController/files/14899891/replay.1.zip)

**Reply by tagyoureit** [reply]
> SpaREMChem says it's chlorinating, but Spa IC20 says output 0%, and the IC20 physical panel shows 0%.

In the logs, the ORP is in the mixing state for both chlorinators and remains that way.  Is the display showing that the chlor was chlorinating during this replay?  It looks like it is off per the poolState.json.

```
[4/7/2024, 9:44:42 PM] verbose: Chem mixing orp remaining: 4min 22sec
...
[4/7/2024, 9:44:43 PM] verbose: Chem mixing orp remaining: 1hr 24min 43sec
```

> When I turn off the pool IC40 via relay, SpaRemChem reports a "chlorinator comms error" for the spa IC20. When I turn the pool IC40 back on, the "chlorinator comms error" on the spa goes away. However, on my connections tab, the spa IC20 reports 0 failures.

I'm also not seeing any of the circuits toggled in this replay.  There is only one send failure in the comms in the entire log.
When you do turn the circuit off, or just in general, are you seeing other stats increase on the same port?  

I can also tell you that the alarms are system wide.  You won't get a different alarm for the IC40 vs the IC20.  It would take quite a bit of rework to do alarms at the equipment level.  If you turn off the IC20 instead, you should see the same alarm.

**Reply by TopJim** [reply]
Thx for looking--good to know about the alarms. I must have sent the wrong replay. I'll try again tonight, sorry about that.

**Reply by TopJim** [reply]
I couldn’t make my ORP probes (lab grade) act right so I ordered the industrial ones, see how it goes, then I’ll re-attack my config issue, examine the port traffic, send a new replay, if necessary.

**Comment by TopJim** [comment]
Problem: "Dynamic" is not controlling my Chlorinators
Re-attacking this issue, now that I have reliable ORP probes. I set dynamic on the IC20 and IC40, but they do not appear to be controlled by NJS-PC. On DashPanel and on the physical IC20 and IC40, they all say 1% which is what I had set it to before I turned on Dynamic. Also there is a "Chlorinator Comms Error". Prior to turning on dynamic, I controlled them manually in DashPanel with no problems. I switch back to manual and the chlorinator comms error disappears and I can control the chlorinators manually, as before.
My setup is NJS-PC alongside i10D.
Intellicenter on ttyUSB2, device 9, AB6FLM99 (Primary)
IC20 Spa on ttyUSB1, device 8, ABSCDHNL (Aux1)
IC40 Pool on ttyUSB0, device 6, A940TBGZ (Aux2)
Spa is on RemChem Controller 51
Pool is on RemChem controller 50
[replay (1).zip](https://github.com/user-attachments/files/17438423/replay.1.zip)

**Reply by tagyoureit** [reply]
I did some work to support multiple chlorinators.  Can you pull a new branch for both njsPC and dp?  In each respective directory you'll need to switch to the new branch and then pull the changes:

1. change to each respective directory and repeat the following
2. `git checkout dual-remchem-dual-chlor`
3. `git pull`

then restart both apps.  You will need to force refresh the browser page (control-refresh) and will see a new drop-down for the chlorinator in each rem chem.  Set it for both controllers.

![image](https://github.com/user-attachments/assets/18663082-488c-4143-abfa-c68fd6b128c6)

**Comment by TopJim** [comment]
Thanks. I’ll get to it as soon as I can, probably on the 29th or 30th as I’ll be at work until then. Davey JonesOn Dec 25, 2024, at 21:19, tagyoureit ***@***.***> wrote:﻿
I did some work to support multiple chlorinators.  Can you pull a new branch for both njsPC and dp?  In each respective directory you'll need to switch to the new branch and then pull the changes:

change to each respective directory and repeat the following
git checkout dual-remchem-dual-chlor
git pull

then restart both apps.  You will need to force refresh the browser page (control-refresh) and will see a new drop-down for the chlorinator in each rem chem.  Set it for both controllers.
image.png (view on web)

**Comment by TopJim** [comment]
I tried the 1st command as you typed it and it says “error: pathspec ‘dual-remchem-dual-chlor’ did not match any files known to git”Davey JonesOn Dec 25, 2024, at 21:19, tagyoureit ***@***.***> wrote:﻿
I did some work to support multiple chlorinators.  Can you pull a new branch for both njsPC and dp?  In each respective directory you'll need to switch to the new branch and then pull the changes:

change to each respective directory and repeat the following
git checkout dual-remchem-dual-chlor
git pull

then restart both apps.  You will need to force refresh the browser page (control-refresh) and will see a new drop-down for the chlorinator in each rem chem.  Set it for both controllers.
image.png (view on web)

**Reply by TopJim** [reply, marked as answer]
Nevermind. I figured it out and just did a git pull. Going thru the dynamic dosing now. Seems to work great, thanks for fixing it! I'll continue to monitor. I kept both bodies open this year, so I can verify that it's dynamicallyt controlling both IC20 and IC40 now.

**Reply by tagyoureit** [reply]
I pushed this up as v8.1.0.  You can do `git checkout master` to switch back to the main branch and then pull the changes.

**Reply by TopJim** [reply]
Done. Thanks again for the revision. I've been running Dynamic on the IC20 and IC40 since 26 Dec. Tested the Free Chlorine today and it's doing well: pool 3.5; spa 2.
Before I had dynamic available to me, I was running the spa IC20 at 0.5% (1% at 12 hrs per day) and then doing a superchlorinate for 1 hr whenever I used the spa. For the pool I did 4% when not in use and 40% when in use, although I found that in the winter it was only requiring 1%.
This dynamic is holding the parameters a lot better than any of my previous methods. I'm using the Sensorex Gold ORP probe (lab edition) on the pool and the Atlas Gold ORP (lab edition) on the Spa. I noticed about a 50 mV variation between the two with the Sensorex being much less stable than the Atlas (Sensorex tends to read 700 and Atlas 750 for a FC of about 2, with both at about 7.5 pH). So I'll go with Atlas lab grade gold ORP for the pool also whenever the Sensorex needs replacement.
